spec-tests/src/test/java/org/candlepin/spec/CloudRegistrationSpecTest.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2009 - 2022 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.spec;
+
+import static org.candlepin.spec.bootstrap.assertions.StatusCodeAssertions.assertBadRequest;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.candlepin.ApiException;
+import org.candlepin.dto.api.v1.CloudRegistrationDTO;
+import org.candlepin.dto.api.v1.ConsumerDTO;
+import org.candlepin.dto.api.v1.OwnerDTO;
+import org.candlepin.resource.CloudRegistrationApi;
+import org.candlepin.spec.bootstrap.assertions.OnlyInHosted;
+import org.candlepin.spec.bootstrap.client.ApiClient;
+import org.candlepin.spec.bootstrap.client.ApiClients;
+import org.candlepin.spec.bootstrap.client.SpecTest;
+import org.candlepin.spec.bootstrap.data.builder.Consumers;
+import org.candlepin.spec.bootstrap.data.builder.Owners;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.Set;
+
+
+@SpecTest
+public class CloudRegistrationSpecTest {
+
+    private static ApiClient client;
+    private static OwnerDTO owner;
+    private CloudRegistrationDTO token;
+
+
+    //TODO Here we should use upstream owner
+    private CloudRegistrationDTO generateToken(OwnerDTO owner, String type, String signature) {
+        CloudRegistrationDTO registrationDTO = new CloudRegistrationDTO();
+        registrationDTO.setType(type == null ? "test_type" : type);
+        registrationDTO.setMetadata(owner.getKey());
+        registrationDTO.setSignature(signature == null ? "test_signature" : signature);
+        return registrationDTO;
+    }
+
+    @BeforeAll
+    static void beforeAll() throws ApiException {
+        client = ApiClients.admin();
```

Comment:
The admin client might be stateless, but the other clients most certainly are not, as they're tied to a specific user, owner, consumer, or cert. And in the admin client case, there should be almost no resource utilization behind spinning it up, as it should be little more than a set of wrappers around serialization and TCP comms.

However, the my larger concern is:
> a test should be fully isolated.

The more we throw stuff into a `beforeEach` or `beforeAll`, the more tests get chunky, flaky, and annoying to maintain, especially in some cases where we want or expect the state to be pristine, but have inherited stuff from the global setup. In fact, someone is going to run into exactly this problem while we port some of the spec tests, as there is at least one suite of tests which have the "wrong" count of objects for the tests because of inherited state from a `before(:each)` block.

It's a paradigm we should avoid, and discourage over callable setup/utility methods where possible and appropriate.

Reply:
I agree with you. `beforeAll/beforeEach` should be kept minimal. Big setup block only make reasoning about the tests harder. There are of course cases where some share setup is justified. Like those clients. They are immutable and thread safe. We gain nothing but inconvenience by inlining them into tests.

Reply:
done